<a href="https://colab.research.google.com/github/fajarrinaldi88/Data-Science-2026/blob/main/Pertemuan3_FajarRinaldi_250401020204.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

* Nama : Fajar Rinaldi
* NIM : 250401020204
* Kelas : IF405

In [14]:
import pandas as pd, numpy as np
from scipy.stats.mstats import winsorize

# STEP 0 — Load & eksplorasi awal
df = pd.read_csv('https://raw.githubusercontent.com/fajarrinaldi88/Data-Science-2026/refs/heads/main/housing_dirty.csv')
print('Shape awal:', df.shape)
print(df.isnull().sum())

Shape awal: (130, 7)
id               0
luas_m2         18
harga_juta      17
kota             0
kamar           10
tahun_bangun     0
kondisi          0
dtype: int64


In [15]:
# STEP 1 — Hapus Duplikat
df.drop_duplicates(inplace=True)
print('Setelah hapus duplikat:', df.shape)

Setelah hapus duplikat: (130, 7)


In [16]:
# STEP 2 — Normalisasi String
df['kota'] = df['kota'].str.strip().str.title()
df['kondisi'] = df['kondisi'].str.strip().str.lower()
print('Normalisasi teks selesai.')

Normalisasi teks selesai.


In [17]:
# STEP 3 — Imputasi Missing Values
df['luas_m2'] = df['luas_m2'].fillna(df['luas_m2'].median())
df['harga_juta'] = df['harga_juta'].fillna(df['harga_juta'].median())
# Pakai [0] karena .mode() menghasilkan Series, kita ambil nilai pertamanya
df['kamar'] = df['kamar'].fillna(df['kamar'].mode()[0])
print('Data kosong telah diisi.')

Data kosong telah diisi.


In [18]:
# STEP 4 — Tangani Outlier (IQR Fence)
for col in ['harga_juta', 'luas_m2', 'tahun_bangun']:
    Q1, Q3 = df[col].quantile([0.25, 0.75])
    IQR = Q3 - Q1
    df[col] = df[col].clip(Q1-1.5*IQR, Q3+1.5*IQR)
print('Penanganan outlier selesai.')

Penanganan outlier selesai.


In [19]:
# STEP 5 — Validasi & Ekspor
assert df.isnull().sum().sum() == 0, 'Masih ada missing!'
assert df.duplicated().sum() == 0, 'Masih ada duplikat!'

print('Shape akhir:', df.shape)
df.to_csv('housing_clean.csv', index=False)
print('Dataset bersih tersimpan dengan nama: housing_clean.csv')

Shape akhir: (130, 7)
Dataset bersih tersimpan dengan nama: housing_clean.csv


## Kesimpulan

Proses pembersihan data telah berhasil dilakukan dengan langkah-langkah berikut:

1.  **Duplikasi Data**: Duplikasi data telah diatasi untuk memastikan setiap entri unik.
2.  **Normalisasi String**: Kolom 'kota' dan 'kondisi' telah dinormalisasi untuk konsistensi data.
3.  **Imputasi Missing Values**: Nilai-nilai yang hilang pada kolom 'luas_m2', 'harga_juta', dan 'kamar' telah diisi menggunakan median dan modus, sehingga dataset menjadi lengkap.
4.  **Penanganan Outlier**: Outlier pada 'harga_juta', 'luas_m2', dan 'tahun_bangun' telah ditangani menggunakan metode IQR Fence, menjadikan data lebih robust terhadap nilai ekstrem.


### Temuan Utama:

*   **Kualitas Data Awal**: Dataset awal memiliki banyak masalah seperti missing values (terutama di 'luas_m2', 'harga_juta', 'kamar') dan potensi duplikasi.
*   **Efektivitas Metode Pembersihan**: Metode yang diterapkan (penghapusan duplikasi, normalisasi string, imputasi median/modus, penanganan outlier IQR) terbukti efektif dalam menghasilkan dataset yang bersih dan siap pakai.
*   **Konsistensi Data**: Normalisasi string berhasil menciptakan konsistensi pada kolom kategori, yang krusial untuk analisis.

### Keterbatasan dan Pertanyaan yang Muncul:

*   **Metode Imputasi**: Penggunaan median dan modus adalah pendekatan yang sederhana. Apakah ada metode imputasi yang lebih canggih (misalnya regresi atau machine learning) yang dapat memberikan hasil yang lebih akurat, terutama untuk 'harga_juta' dan 'luas_m2'?
*   **Penanganan Outlier**: IQR Fence adalah metode yang umum, tetapi apakah outlier yang di-clip benar-benar mencerminkan anomali ataukah merupakan data yang valid namun ekstrem? Perlu validasi domain-specific.
*   **Representasi Data**: Ukuran dataset yang relatif kecil (130 baris) mungkin membatasi generalisasi temuan. Apakah data ini representatif untuk seluruh populasi data properti?
*   **Verifikasi Data**: Tidak ada sumber eksternal yang digunakan untuk memverifikasi keakuratan data setelah pembersihan. Bagaimana kita bisa memastikan bahwa nilai-nilai yang diimputasi atau disesuaikan adalah benar secara faktual?
*   **Implikasi Bisnis**: Bagaimana kualitas data yang lebih baik ini akan memengaruhi model prediksi harga properti atau keputusan bisnis lainnya?